templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,871 @@
+
+    # Adjust batch size per device (BS) and number of devices (ND) according to model size
+    # Number of devices (ND) is determined by a combination of factors, context length, accelerator type, whether LoRA is used, etc.
+    if SIZE == "7b" or SIZE == "13b":  
+        # ND is et on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
+        # If Nvidia A100 is used. ND can be set to 8 instead.
+        BS, ND = 16, 16
+    elif SIZE == "70b":
+        # ND is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
+        # If Nvidia A100 is used. ND can be set to 16 instead.
+        BS, ND = 8, 32
+    else:
+        print(f"Invalid size: {SIZE}")
+        sys.exit(1)
+
+    # Pick model ID and deepspeed config based on model size
+    MODEL_ID = f"meta-llama/Llama-2-{SIZE}-hf"
+    CONFIG_DIR = f"./deepspeed_configs/zero_3_llama_2_{SIZE}.json"
+
+    # Update the config with the adjusted parameters
+    config.update(
+        **{
+            "as_test": args.as_test,
+            "batch_size": BS,
+            "batch_size_per_device": args.batch_size_per_device,
+            "block_size": args.ctx_len,
+            "ds_config": CONFIG_DIR,
+            "eval_batch_size": args.eval_batch_size_per_device,
+            "grad_accum": args.grad_accum,
+            "lora": args.lora,
+            "lr": LR,
+            "mixed_precision": args.mixed_precision,
+            "model_name": MODEL_ID,
+            "no_grad_ckpt": args.no_grad_ckpt,
+            "num_checkpoints_to_keep": args.num_checkpoints_to_keep,
+            "num_devices": ND,
+            "num_epochs": args.num_epochs,
+            "output_dir": args.output_dir,
+            "seed": 42,
+            "special_token_path": args.special_token_path,
+            "stop_perplexity": args.stop_perplexity,
+            "train_path": args.train_path,
+            "test_path": args.test_path,
+        }
+    )
+
+    # Add LoRA config if needed
+    if config["lora"]:
+        with open("./lora_configs/lora.json", "r") as json_file:
+            lora_config = json.load(json_file)
+        config["lora_config"] = lora_config
+
+    # Add deepspeed plugin to the config
+    ds_plugin = DeepSpeedPlugin(hf_ds_config=config.get("ds_config"))
+    config.update(ds_plugin=ds_plugin)
+
+    # Set the local checkpoint directory which may get used when the remote cloud storage bucket checkpointing is disabled.
+    os.environ["RAY_AIR_LOCAL_CACHE_DIR"] = config["output_dir"]
+
+    # Set up runtime_env for Ray Train
+    ray.init(
+        runtime_env={
+            "env_vars": {
+                "HF_HOME": "/mnt/local_storage/.cache/huggingface",
+                "RAY_AIR_LOCAL_CACHE_DIR": os.environ["RAY_AIR_LOCAL_CACHE_DIR"],
+            },
+            # This is not needed to run in Workspace. However, it might be needed for other environments.
+            # "working_dir": ".",
```

Comment:
According to our offline discussion, this can go.

Reply:
Done